File: packages/exif/ios/Plugin/Classes/Options/WriteExifOptions.swift
import Foundation
import Capacitor

@objc public class WriteExifOptions: NSObject {
    let dateTimeOriginal: String?
    let exposureTime: Double?
    let fNumber: Double?
    let flash: Bool?
    let focalLength: Double?
    let gpsAltitude: Double?
    let gpsLatitude: Double?
    let gpsLongitude: Double?
    let iso: Int?
    let lensModel: String?
    let make: String?
    let model: String?
    let orientation: Int?
    let software: String?
    let url: URL

    init(_ call: CAPPluginCall) throws {
        let tags = try WriteExifOptions.getTagsFromCall(call)
        self.dateTimeOriginal = tags["dateTimeOriginal"] as? String
        self.exposureTime = (tags["exposureTime"] as? NSNumber)?.doubleValue
        self.fNumber = (tags["fNumber"] as? NSNumber)?.doubleValue
        self.flash = tags["flash"] as? Bool
        self.focalLength = (tags["focalLength"] as? NSNumber)?.doubleValue
        self.gpsAltitude = (tags["gpsAltitude"] as? NSNumber)?.doubleValue
        self.gpsLatitude = (tags["gpsLatitude"] as? NSNumber)?.doubleValue
        self.gpsLongitude = (tags["gpsLongitude"] as? NSNumber)?.doubleValue
        self.iso = (tags["iso"] as? NSNumber)?.intValue
        self.lensModel = tags["lensModel"] as? String
        self.make = tags["make"] as? String
        self.model = tags["model"] as? String
        self.orientation = (tags["orientation"] as? NSNumber)?.intValue
        self.software = tags["software"] as? String
        self.url = try ReadExifOptions.getUrlFromCall(call)
        if (gpsLatitude == nil) != (gpsLongitude == nil) {
            throw CustomError.gpsCoordinatesIncomplete
        }
    }

    private static func getTagsFromCall(_ call: CAPPluginCall) throws -> JSObject {
        guard let tags = call.getObject("tags") else {
            throw CustomError.tagsMissing
        }
        return tags
    }
}

File: packages/exif/android/src/main/java/io/capawesome/capacitorjs/plugins/exif/classes/options/WriteExifOptions.java
package io.capawesome.capacitorjs.plugins.exif.classes.options;

import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import io.capawesome.capacitorjs.plugins.exif.classes.CustomExceptions;

public class WriteExifOptions {

    @Nullable
    private final String dateTimeOriginal;

    @Nullable
    private final Double exposureTime;

    @Nullable
    private final Double fNumber;

    @Nullable
    private final Boolean flash;

    @Nullable
    private final Double focalLength;

    @Nullable
    private final Double gpsAltitude;

    @Nullable
    private final Double gpsLatitude;

    @Nullable
    private final Double gpsLongitude;

    @Nullable
    private final Integer iso;

    @Nullable
    private final String lensModel;

    @Nullable
    private final String make;

    @Nullable
    private final String model;

    @Nullable
    private final Integer orientation;

    @NonNull
    private final String path;

    @Nullable
    private final String software;

    public WriteExifOptions(@NonNull PluginCall call) throws Exception {
        JSObject tags = WriteExifOptions.getTagsFromCall(call);
        this.dateTimeOriginal = tags.getString("dateTimeOriginal");
        this.exposureTime = WriteExifOptions.getDoubleFromTags(tags, "exposureTime");
        this.fNumber = WriteExifOptions.getDoubleFromTags(tags, "fNumber");
        this.flash = tags.getBool("flash");
        this.focalLength = WriteExifOptions.getDoubleFromTags(tags, "focalLength");
        this.gpsAltitude = WriteExifOptions.getDoubleFromTags(tags, "gpsAltitude");
        this.gpsLatitude = WriteExifOptions.getDoubleFromTags(tags, "gpsLatitude");
        this.gpsLongitude = WriteExifOptions.getDoubleFromTags(tags, "gpsLongitude");
        this.iso = tags.getInteger("iso");
        this.lensModel = tags.getString("lensModel");
        this.make = tags.getString("make");
        this.model = tags.getString("model");
        this.orientation = tags.getInteger("orientation");
        this.path = WriteExifOptions.getPathFromCall(call);
        this.software = tags.getString("software");
        if ((gpsLatitude == null) != (gpsLongitude == null)) {
            throw CustomExceptions.GPS_COORDINATES_INCOMPLETE;
        }
    }

    @Nullable
    public String getDateTimeOriginal() {
        return dateTimeOriginal;
    }

    @Nullable
    public Double getExposureTime() {
        return exposureTime;
    }

    @Nullable
    public Double getFNumber() {
        return fNumber;
    }

    @Nullable
    public Boolean getFlash() {
        return flash;
    }

    @Nullable
    public Double getFocalLength() {
        return focalLength;
    }

    @Nullable
    public Double getGpsAltitude() {
        return gpsAltitude;
    }

    @Nullable
    public Double getGpsLatitude() {
        return gpsLatitude;
    }

    @Nullable
    public Double getGpsLongitude() {
        return gpsLongitude;
    }

    @Nullable
    public Integer getIso() {
        return iso;
    }

    @Nullable
    public String getLensModel() {
        return lensModel;
    }

    @Nullable
    public String getMake() {
        return make;
    }

    @Nullable
    public String getModel() {
        return model;
    }

    @Nullable
    public Integer getOrientation() {
        return orientation;
    }

    @NonNull
    public String getPath() {
        return path;
    }

    @Nullable
    public String getSoftware() {
        return software;
    }

    @Nullable
    private static Double getDoubleFromTags(@NonNull JSObject tags, @NonNull String key) throws Exception {
        if (!tags.has(key)) {
            return null;
        }
        return tags.getDouble(key);
    }

    @NonNull
    private static String getPathFromCall(@NonNull PluginCall call) throws Exception {
        String path = call.getString("path");
        if (path == null || path.isEmpty()) {
            throw CustomExceptions.PATH_MISSING;
        }
        if (path.startsWith("file://")) {
            String uriPath = Uri.parse(path).getPath();
            if (uriPath != null) {
                return uriPath;
            }
        }
        return path;
    }

    @NonNull
    private static JSObject getTagsFromCall(@NonNull PluginCall call) throws Exception {
        JSObject tags = call.getObject("tags");
        if (tags == null) {
            throw CustomExceptions.TAGS_MISSING;
        }
        return tags;
    }
}
